### PR TITLE
Remote file transfer

### PR DESCRIPTION
**Pull request by tudorlesan**

### CODE REVIEW

**Comment by meirwah** [comment]
Do we support KeyFile also?

File: content/org/openscore/slang/base/remote_command_execution/remote_file_transfer/ScpTo.sl [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+# Results:
+#    - SUCCESS - the SSH access was successful and returned with code 0
+#    - FAILURE
+###############################################################################################################################################################################
+
+namespace: org.openscore.slang.base.remote_command_execution.remote_file_transfer
+
+operation:
+    name: ScpTo
+    action:
```
Comment (meirwah) [inline content/org/openscore/slang/base/remote_command_execution/remote_file_transfer/ScpTo.sl]: i think action should be after inputs no?

**Comment by orius123** [comment]
Just from a brief overview.. lower case names and underscores.

File: content/org/openscore/slang/base/remote_command_execution/remote_file_transfer/ScpTo.sl (hunk 1 above)
Comment (tudorlesan) [inline content/org/openscore/slang/base/remote_command_execution/remote_file_transfer/ScpTo.sl]: Sorry, I didn't proofread the operation. I just checked it works. I think that's why I didn't make a pull request

**Comment by meirwah** [comment]
ScpTo is not a good name.. how is this operation is called at OO content?

**Comment by tudorlesan** [comment]
Not yet. I will add it

File: content/org/openscore/slang/base/remote_command_execution/remote_file_transfer/ScpTo.sl (hunk 1 above)
Comment (meirwah) [inline content/org/openscore/slang/base/remote_command_execution/remote_file_transfer/ScpTo.sl]: This is a Pull Request  #50 ... you needed to keep it a branch in that case...

**Comment by tudorlesan** [comment]
@orius123  I thought we agreed that the inputs/outputs for Java actions would remain camelcase and only for python the underscores. I can change them if you want. @Meir It's called Remote Secure Copy but it's more complex than this one, that's why I gave it a different name. I can put the same name

**Comment by meirwah** [comment]
I prefer Remote Secure Copy , then ScpTo ... @orius123  WDYT?

File: content/org/openscore/slang/base/remote_command_execution/remote_file_transfer/ScpTo.sl (hunk 1 above)
Comment (tudorlesan) [inline content/org/openscore/slang/base/remote_command_execution/remote_file_transfer/ScpTo.sl]: ok, sorry about that. Made it to early then.

**Comment by orius123** [comment]
I'm fine with camel case for inputs but op name and outputs should be underscored IMHO.
What about copy remote file?

**Comment by tudorlesan** [comment]
@orius123 ok, but they would be different than SSH and mail operations from score-actions. If there's no problem with that I will change them. I recall using something like - return_result: returnResult as outputs for the openstack operations, so that at least at a superior level, the outputs are underscored.

**Comment by orius123** [comment]
I think that whoever consumes this operation should get lowercase outputs. You can bind directly or with locals or any other method.

**Comment by oritstone** [comment]
@tudorlesan due to the renaming effort we are going to start next week, let's put this issue on hold for now

**Comment by tudorlesan** [comment]
@oritstone it's already on hold since I'm currently working content output changes

**Comment by tudorlesan** [comment]
[ERROR] - Test case failed: testDocker_start_container_Success - Tests that start containers works
		Expected result: SUCCESS
		Actual result: MACHINE_IS_NOT_CLEAN
[ERROR] - Test case failed: testDocker_stop_container_Success - Tests that stop containers works
		Expected result: SUCCESS
		Actual result: PREREQUST_MACHINE_IS_NOT_CLEAN
[ERROR] - Test case failed: test_docker_get_container_names - Tests SUCCESS case
		Expected result: SUCCESS
		Actual result: CONTAINER_NAMES_VERIFY_PROBLEM
[ERROR] - Test case failed: testDocker_delete_container_Success - Tests that delete containers works
		Expected result: SUCCESS
		Actual result: PREREQUISITE_MACHINE_IS_NOT_CLEAN
[ERROR] - Test case failed: testDocker_run_container_Success - Tests that run containers works
		Expected result: SUCCESS
		Actual result: PREREQUST_MACHINE_IS_NOT_CLEAN
[ERROR] - Test case failed: test_get_filtered_containers_success - Tests SUCCESS case - filters out containers based on busybox:latest image
		Expected result: SUCCESS
		Actual result: POST_CLEAR_MACHINE_PROBLEM

**Comment by tudorlesan** [comment]
@orius123 any suggestion on how to solve these?

**Comment by Bonczidai** [comment]
first we need to understand what causes the problem -  it is more problematic since we don't have builder log yet (#250); not so elegant way is to add debug steps to the flows

p.s. these errors are sporadic and happen on multiple PRs, right? maybe they deserve an issue

**Comment by orius123** [comment]
Indeed. But I don't really have a solution ATM

**Comment by tudorlesan** [comment]
tested KeyFile :+1:

**Comment by orius123** [comment]
I think 'src' and 'dest' are better for inputs names

**Comment by tudorlesan** [comment]
@meirwah can you review it please?

File: content/io/cloudslang/base/remote_command_execution/remote_file_transfer/remote_secure_copy.sl [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+#
+#   The Apache License is available at
+#   http://www.apache.org/licenses/LICENSE-2.0
+
+###############################################################################################################################################################################
+#   Copies a file from the local machine to a remote machine or from a remote machine to a different remote machine using the SCP protocol.
+#
+#   Inputs:
+#       - srcHost - optional - host of the source machine (only if remote to remote)
+#       - srcPath - path of the file about to be copied
```
Comment (meirwah) [inline content/io/cloudslang/base/remote_command_execution/remote_file_transfer/remote_secure_copy.sl]: absolute path or relative? or both?

File: test/io/cloudslang/base/remote_command_execution/remote_file_transfer/test_remote_secure_copy.sl [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+          FAILURE: KEY_ADDITION_FAIL
+
+    - encrypt_and_store_authorized_keys:
+        do:
+          base_cmd.run_command:
+            - command: "'AUTHORIZED_KEYS=$(base64 -w0 ~/.ssh/authorized_keys)'"
+        navigate:
+          SUCCESS: create_needed_folders
+          FAILURE: KEY_STORE_FAIL
+
```
Comment (meirwah) [inline test/io/cloudslang/base/remote_command_execution/remote_file_transfer/test_remote_secure_copy.sl]: steps : generate_key + add_key_to_authorized + add_key_to_authorized + encrypt_and_store_authorized_keys should be moved to a subflow... and be add to the content (not the test content)

File: test/io/cloudslang/base/remote_command_execution/remote_file_transfer/test_remote_secure_copy.sl [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+        navigate:
+          SUCCESS: test_remote_secure_copy
+          FAILURE: FILE_REACHING_SRC_HOST_FAIL
+
+    - test_remote_secure_copy:
+        do:
+          base_rft.remote_secure_copy:
+            - sourceHost: "'localhost'"
+            - sourcePath: "'/home/data/file.txt'"
+            - sourcePort: "'22022'"
```
Comment (meirwah) [inline test/io/cloudslang/base/remote_command_execution/remote_file_transfer/test_remote_secure_copy.sl]: why hard coded '22022'? why not use flow vars?

File: test/io/cloudslang/base/remote_command_execution/remote_file_transfer/remote_secure_copy.inputs.yaml [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+####################################################
+testRemoteSecureCopySucess:
+  inputs:
+    - host: localhost
+    - port: "49153"
+    - username: root
+    - password: screencast
+  description: Tests rft successfully transfers a file from one remote to another
```
Comment (meirwah) [inline test/io/cloudslang/base/remote_command_execution/remote_file_transfer/remote_secure_copy.inputs.yaml]: where is a test from local to remote?

File: content/io/cloudslang/base/remote_command_execution/remote_file_transfer/remote_secure_copy.sl (hunk 2 above)
Comment (tudorlesan) [inline content/io/cloudslang/base/remote_command_execution/remote_file_transfer/remote_secure_copy.sl]: both

Comment (meirwah) [inline content/io/cloudslang/base/remote_command_execution/remote_file_transfer/remote_secure_copy.sl]: write it down...

File: content/io/cloudslang/base/remote_command_execution/remote_file_transfer/remote_secure_copy.sl [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+###############################################################################################################################################################################
+#   Copies a file from the local machine to a remote machine or from a remote machine to a different remote machine using the SCP protocol.
+#
+#   Inputs:
+#       - srcHost - optional - host of the source machine (only if remote to remote)
+#       - srcPath - absolute or relative path of the file about to be copied
+#       - srcUsername - optional - username of the source machine (only if remote to remote)
+#       - srcPassword - optional -  password of the source machine (only if remote to remote)
+#       - srcPort - optional - port number for the source machine (only if remote to remote) - Default: 22
+#       - srcPrivateKeyFile - optional - path to the private key file on the source machine (only if remote to remote)
```
Comment (Bonczidai) [inline content/io/cloudslang/base/remote_command_execution/remote_file_transfer/remote_secure_copy.sl]: for all the inputs which start with `src` -> `source` (according to inputs section)

File: content/io/cloudslang/base/remote_command_execution/remote_file_transfer/remote_secure_copy.sl [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+#       - destHost - host of the destination machine
+#       - destPath - absolute or relative path where the file will be copied
+#       - destPort - optional - port number for the destination machine - Default: 22
+#       - destUsername - username of the destination machine
+#       - destPassword - optional - password of the destination machine
+#       - destPrivateKeyFile optional - path to the private key file on the destination machine
```
Comment (Bonczidai) [inline content/io/cloudslang/base/remote_command_execution/remote_file_transfer/remote_secure_copy.sl]: same for `destination`

File: test/io/cloudslang/base/remote_command_execution/remote_file_transfer/remote_secure_copy.inputs.yaml [hunk 8]
```diff
@@ -0,0 +1,42 @@
+#   (c) Copyright 2014 Hewlett-Packard Development Company, L.P.
+#   All rights reserved. This program and the accompanying materials
+#   are made available under the terms of the Apache License v2.0 which accompany this distribution.
+#
+#   The Apache License is available at
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+####################################################
+testRemoteSecureCopySucess:
```
Comment (Bonczidai) [inline test/io/cloudslang/base/remote_command_execution/remote_file_transfer/remote_secure_copy.inputs.yaml]: maybe rename this to `...SucessRemoteToRemote` like the second test

File: test/io/cloudslang/base/remote_command_execution/remote_file_transfer/test_remote_secure_copy.sl [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+    - check_text:
+        do:
+          base_strings.string_equals:
+            - first_string: text_to_check
+            - second_string: read_text
+        navigate:
+          SUCCESS: post_clear_machine
+          FAILURE: FILE_CHECK_FAIL
+
+    - post_clear_machine:
```
Comment (Bonczidai) [inline test/io/cloudslang/base/remote_command_execution/remote_file_transfer/test_remote_secure_copy.sl]: please check if you need this task

File: test/io/cloudslang/base/remote_command_execution/remote_file_transfer/test_remote_secure_copy.sl [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+                required: false
+            - port:
+                required: false
+        navigate:
+          SUCCESS: SUCCESS
+          FAILURE: POSTREQUEST_MACHINE_IS_NOT_CLEAN
+
+  results:
+    - SUCCESS
+    - FAILURE
```
Comment (Bonczidai) [inline test/io/cloudslang/base/remote_command_execution/remote_file_transfer/test_remote_secure_copy.sl]: please add custom failure result here

File: test/io/cloudslang/base/remote_command_execution/remote_file_transfer/remote_secure_copy.inputs.yaml (hunk 8 above)
Comment (Bonczidai) [inline test/io/cloudslang/base/remote_command_execution/remote_file_transfer/remote_secure_copy.inputs.yaml]: if you are going to proceeed with this, then also rename the test flow

File: test/io/cloudslang/base/remote_command_execution/remote_file_transfer/test_remote_secure_copy_local_to_remote.sl [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+        do:
+          base_strings.string_equals:
+            - first_string: text_to_check
+            - second_string: read_text
+        navigate:
+          SUCCESS: post_clear_machine
+          FAILURE: FILE_CHECK_FAIL
+
+    - post_clear_machine:
+        do:
```
Comment (Bonczidai) [inline test/io/cloudslang/base/remote_command_execution/remote_file_transfer/test_remote_secure_copy_local_to_remote.sl]: please check if you need this task

File: test/io/cloudslang/base/remote_command_execution/remote_file_transfer/test_remote_secure_copy_local_to_remote.sl [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+                required: false
+            - port:
+                required: false
+        navigate:
+          SUCCESS: SUCCESS
+          FAILURE: POSTREQUEST_MACHINE_IS_NOT_CLEAN
+
+  results:
+    - SUCCESS
+    - FAILURE
```
Comment (Bonczidai) [inline test/io/cloudslang/base/remote_command_execution/remote_file_transfer/test_remote_secure_copy_local_to_remote.sl]: please add custom failure result here

File: test/io/cloudslang/base/remote_command_execution/remote_file_transfer/remote_secure_copy.inputs.yaml [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+    - password: screencast
+    - scp_host_port: "22022"
+    - key_name: key
+    - scp_username: data
+    - scp_path: /home/data/
+    - scp_file: file.txt
+    - text_to_check: "hello world"
+  description: Tests rft successfully transfers a file from local to another
+  testFlowPath: io.cloudslang.base.remote_command_execution.remote_file_transfer.test_remote_secure_copy_local_to_remote
+  testSuites: [docker,docker-misc]
```
Comment (Bonczidai) [inline test/io/cloudslang/base/remote_command_execution/remote_file_transfer/remote_secure_copy.inputs.yaml]: I think you should create a new test suite for `remote-copy` and do something similar:

```
 matrix:
   - SUITE=containers
   - SUITE=images
   - SUITE=docker-misc
   - SUITE=git,remote-copy
```

**Comment by Bonczidai** [comment]
added my comments, looks good

**Comment by mihaitusa** [comment]
@gituser173  Looks good. :+1:

**Pull request merged**